File: binary-tree-level-order-traversal - DFS.py
# Time complexity: O(n) where n is the number of nodes
# Space complexity: O(h) where h is the height of the Tree

# The code successfully ran on Leetcode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.result = []
        
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root == None: return self.result
        
        self.dfs(root, 0)
        
        return self.result
    
    def dfs(self, root, lvl):
        # Base
        if root == None: return
        
        # Logic
        if len(self.result)== lvl:
            self.result.append([])
            
        self.result[lvl].append(root.val)
        
        self.dfs(root.left, lvl + 1)
        self.dfs(root.right, lvl + 1)

File: binary-tree-level-order-traversal-dfs.cpp
// Time Complexity : O(n) total nodes
// Space Complexity : O(m) max nodes
// Did this code successfully run on Leetcode : yes

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a binary tree node
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
private:
    vector<vector<int>> result;
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        helper(root, 0);
        return result;
    }
    void helper(TreeNode* root, int level){
        if(root == nullptr) return;

        if(level == result.size()){
            result.push_back({root->val});
        }
        else{
            result[level].push_back(root->val);
        }

        helper(root->left, level+1);
        helper(root->right, level+1);
    }
};